[charts] Use new text component to avoid tick label overflow on x-axis (#10648)
mui/mui-x: docs/data/charts/axis/axis.md

{{"demo": "TickNumber.js"}}

### Fix tick positions

If you want more control on the tick positions, you can use the `tickInterval` property.

> This property accept an array of values.

Reviewer: Then maybe in the next sentence mention that only for this scale type the property can ALSO be a callback..? 🤔

Ticks will be placed at those values.

For axis with scale type `'point'`, the `tickInterval` property can be a filtering function of the type `(value, index) => boolean`.

In the next demo, both axes are with `scaleType='point'`.
The top axis is here to display the default behavior.
It shows a tick for each point.
The bottom axis use a filtering function to only display a tick at the beginning of a day.

{{"demo": "TickPosition.js"}}

### Filtering ticks label

You can display labels only on a subsets of the ticks with the `tickInterval` property.
It's a filtering function `(value, index) => boolean`.

> For example `tickInterval: (value, index) => index % 2 === 0` will show only one label every two ticks.

Author: ```suggestion
For example `tickLabelInterval: (value, index) => index % 2 === 0` will show the label only on one tick out of two.
```

> For example `tickInterval: (value, index) => index % 2 === 0` will show only one label every two ticks.

Author: I agree but "every other something" is an idiomatic expression so either you know it or not. So for a non native speaker (like me 30 minutes ago) this sentence is useless 😉

The [cambride](https://dictionary.cambridge.org/fr/dictionnaire/anglais/every-other) definition is "not each one in a series, but every two". SO maybe this is a correct and non idiomatic version

```suggestion
For example `tickLabelInterval: (value, index) => index % 2 === 0` will show the label every two ticks.
```

> For example `tickInterval: (value, index) => index % 2 === 0` will show only one label every two ticks.

Reviewer: Fair point and the latest suggestion does make sense to me as well. 👍 💯

:::warning
The `value` and `index` in the argument are those of the ticks, not the axis data.
:::

By default, ticks are filtered such that their labels do not overlap.
You can override this behavior with `tickInterval: () => true` which show the tick label for each tick.

> In this example we still have the top axis as a reference for the default behavior.

Reviewer: ```suggestion
In this example, the top axis is a reference for the default behavior.
```

{{"demo": "AxisCustomizationNoSnap.js", "hideToolbar": true, "bg": "inline"}}

> ### Text Customization

Reviewer: ```suggestion
### Text customization
```

To customize the text elements (ticks label and the axis label) use the `tickLabelStyle` and `labelStyle` properties of the axis configuration.

> {{"demo": "AxisTextCustomizationNoSnap.js", "hideToolbar": true, "bg": "inline"}}

Reviewer: WDYT about the fact that the tick labels almost overlap the tick itself with the default angle of 45 degrees and the more you rotate towards 90 degrees, the worse it gets. 🤔 
Should we maybe include some margin? Or the text management itself should account for it? 🤔 

https://github.com/mui/mui-x/assets/4941090/61d33aef-d2b8-43a6-9ae7-4c03826b37b0

Reviewer: Cool, the updated demo looks great! 🙌
